src/Cassandra/Mapping/Statements/StatementFactory.cs

```diff
@@ -39,25 +40,49 @@ public StatementFactory()
                 SetStatementProperties(statement, cql);
                 return statement;
             }
-            var ps = await _statementCache
-                .GetOrAdd(cql.Statement, query => PrepareNew(query, session))
-                .ConfigureAwait(false);
-            var boundStatement = ps.Bind(cql.Arguments);
-            SetStatementProperties(boundStatement, cql);
-            return boundStatement;
-        }
 
-        private Task<PreparedStatement> PrepareNew(string query, ISession session)
-        {
-            var count = Interlocked.Increment(ref _statementCacheCount);
-            if (count > MaxPreparedStatementsThreshold)
+            var wasPreviouslyCached = true;
+
+            var psCacheKey = new CacheKey(cql.Statement, session);
+            var query = cql.Statement;
+
+            var prepareTask = _statementCache.GetOrAdd(psCacheKey, _ =>
+            {
+                wasPreviouslyCached = false;
+                return session.PrepareAsync(query);
+            });
+
+            PreparedStatement ps;
+            try
+            {
+                ps = await prepareTask.ConfigureAwait(false);
+            }
+            catch (Exception) when (wasPreviouslyCached)
             {
-                Logger.Warning("The prepared statement cache contains {0} queries. This issue is probably due " +
-                               "to misuse of the driver, you should use parameter markers for queries. You can " +
-                               "configure this warning threshold using " +
-                               "MappingConfiguration.SetMaxStatementPreparedThreshold() method.", count);
+                // The exception was caused from awaiting upon a Task that was previously cached
+                // It's possible that the schema or topology changed making this query preparation to succeed
+                // in a new attemp
+                prepareTask = session.PrepareAsync(query);
+                ps = await prepareTask.ConfigureAwait(false);
+                // AddOrUpdate() returns a task which we already waited upon, its safe to call Forget()
+                _statementCache.AddOrUpdate(psCacheKey, prepareTask, (k, v) => prepareTask).Forget();
```

Comment:
Why not use `TryUpdate` instead of AddOrUpdate? from what I understand, the key is already there always

Reply:
`TryUpdate()` will only update after a comparison. In this case, I want to update no matter what (last one wins), that is why I use `(k, v) => prepareTask` lambda.